CMakeLists.txt

```diff
@@ -239,6 +250,80 @@ endif()
 
 include(cmake/optional.cmake)
 
+set(hardened_cxx_flags "")
+if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  # All versions of gcc that we commonly use for building are subject to bug
+  # https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90348. To work around that, set
+  # -fstack-reuse=none for all gcc builds.
+  try_append_cxx_flags(hardened_cxx_flags "-fstack-reuse=none")
+endif()
+if(HARDENING)
+  if(MSVC)
+    try_append_cxx_flags(core "/DYNAMICBASE")
+    try_append_cxx_flags(core "/HIGHENTROPYVA")
+    try_append_cxx_flags(core "/NXCOMPAT")
+  else()
+    try_append_cxx_flags(hardened_cxx_flags "-Wstack-protector")
+    try_append_cxx_flags(hardened_cxx_flags "-fstack-protector-all")
+    try_append_cxx_flags(hardened_cxx_flags "-fcf-protection=full")
+    if(MINGW)
+      link_libraries(ssp)
+    else()
+      # stack-clash-protection doesn't currently work, and likely should just be skipped for Windows.
+      # See https://gcc.gnu.org/bugzilla/show_bug.cgi?id=90458 for more details.
+      try_append_cxx_flags(hardened_cxx_flags "-fstack-clash-protection")
+    endif()
+
+    try_append_linker_flag(core "-Wl,--enable-reloc-section")
+    try_append_linker_flag(core "-Wl,--dynamicbase")
+    try_append_linker_flag(core "-Wl,--nxcompat")
+    try_append_linker_flag(core "-Wl,--high-entropy-va")
+    try_append_linker_flag(core "-Wl,-z,relro")
+    try_append_linker_flag(core "-Wl,-z,now")
+    try_append_linker_flag(core "-Wl,-z,separate-code")
+    if(CMAKE_SYSTEM_NAME STREQUAL "Darwin")
+      try_append_linker_flag(core "-Wl,-bind_at_load")
+    endif()
+  endif()
+endif()
+separate_arguments(hardened_cxx_flags)
+target_compile_options(core INTERFACE ${hardened_cxx_flags})
+
+if(REDUCE_EXPORTS)
+  set(CMAKE_CXX_VISIBILITY_PRESET hidden)
+  set(CMAKE_VISIBILITY_INLINES_HIDDEN 1)
+  if(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
```

Comment:
Why is this gated on `GNU`? I could be using clang and ld, and this would work fine (as it does on master)?

Reply:
Thanks! Fixed.